### PR TITLE
feat: [CDS-68278]: OPA policy integration for template

### PR DESCRIPTION
**Pull request by shivamnegi94**
## Describe your changes
Added the OPA support for template service for onsave action
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48336)

### CODE REVIEW

**Comment by shivamnegi94** [comment]
trigger uts

**Review by shalini-agr — CHANGES_REQUESTED** [review]
> Reviewed 7 of 19 files at r1, 12 of 12 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48336)* status: 19 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`template-service/service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1596 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWkbnXmCQy68n7GlEBx:-NWkbnXn4AI7M1urPEm7:bdb559l) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/template-service/service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1596)):*
> ```Java
>     GovernanceMetadata governanceMetadata = governanceService.evaluateGovernancePoliciesForTemplate(
>         templateEntity.getYaml(), templateEntity.getAccountId(), templateEntity.getOrgIdentifier(),
>         templateEntity.getProjectIdentifier(), OpaConstants.OPA_EVALUATION_ACTION_PIPELINE_SAVE);
> ```

let's rename this to OPA_EVALUATION_ACTION_SAVE
___
*[`878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceService.java` line 19 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWka_CFEZdMC6C0Crr2:-NWka_CFEZdMC6C0Crr3:b-vo8qp9) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceService.java#L19)):*
> ```Java
>       String projectIdentifier, String action, String planExecutionId, String pipelineVersion);
>   GovernanceMetadata evaluateGovernancePoliciesForTemplate(
>       String expandedJson, String accountId, String orgIdentifier, String projectIdentifier, String action);
> ```

if we are just planning to send template yaml here, can we rename this variable to templateJson rather than expanded json?
___
*[`878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java` line 104 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWkb7vwC47G65NseOLf:-NWkb7vwC47G65NseOLg:bbe3l15) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java#L104)):*
> ```Java
> 
>   @Override
>   public GovernanceMetadata evaluateGovernancePoliciesForTemplate(
> ```

can we check if we can extract some common code from the similar method for pipeline entity?
we can pass some arguments, like FF name, entity type, etc. to a common method.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48336)* status: 19 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceHelper.java` line 154 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWkikBD4Y4uRUOKPf4q:-NWkikBE4WrS0QEjp-qk:b-mvc88r) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceHelper.java#L154)):*
> ```Java
>                                           .put("entityName", templateName)
>                                           .build();
>     return URLEncoder.encode(JsonUtils.asJson(metadataMap), StandardCharsets.UTF_8.toString());
> ```

Why template service related code needs to be present here?
___
*[`878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java` line 104 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWkb7vwC47G65NseOLf:-NWkisC92gVrLEce1yle:b32v57) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java#L104)):*
<details><summary><i>Previously, shalini-agr (Shalini Agrawal) wrote…</i></summary><blockquote>

can we check if we can extract some common code from the similar method for pipeline entity?
we can pass some arguments, like FF name, entity type, etc. to a common method. 
</blockquote></details>

+1
___
*[`878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java` line 141 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWkiwDR4o-uFi1_NQ_q:-NWkiwDR4o-uFi1_NQ_r:b8x0ox4) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java#L141)):*
> ```Java
>         String entityMetadata =
>             GovernanceServiceHelper.getEntityMetadataStringForTemplate(templateIdentifier, templateName);
>         String userIdentifier = GovernanceServiceHelper.getUserIdentifier();
> ```

Lets try not to use template or pipeline keywords here



_Code quote:_
```Java
      OpaEvaluationResponseHolder response;
      try {
        YamlField pipelineField = YamlUtils.readTree(expandedJson);
        String templateIdentifier =
            pipelineField.getNode().getField(YAMLFieldNameConstants.TEMPLATE).getNode().getIdentifier();
        String templateName = pipelineField.getNode().getField(YAMLFieldNameConstants.TEMPLATE).getNode().getName();
        String entityString = GovernanceServiceHelper.getEntityStringTemplate(
            accountId, orgIdentifier, projectIdentifier, templateIdentifier);
        String entityMetadata =
            GovernanceServiceHelper.getEntityMetadataStringForTemplate(templateIdentifier, templateName);
        String userIdentifier = GovernanceServiceHelper.getUserIdentifier();
```

File: 878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceHelper.java [hunk 1]
```diff
@@ -135,4 +136,28 @@ public String getUserIdentifier() {
     UserPrincipal userPrincipal = (UserPrincipal) SourcePrincipalContextBuilder.getSourcePrincipal();
     return userPrincipal.getName();
   }
+
+  public String getEntityStringTemplate(String accountId, String orgIdentifier, String projectIdentifier,
+      String templateIdentifier) throws UnsupportedEncodingException {
+    String entityStringRaw =
+        String.format("accountIdentifier:%s/orgIdentifier:%s/projectIdentifier:%s/templateIdentifier:%s", accountId,
```
Comment (namanvermah) [inline 878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceHelper.java:143]: Please keep the entity string as per how we discussed with @dwarren-dev

File: 878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceHelper.java [hunk 2]
```diff
@@ -135,4 +136,28 @@ public String getUserIdentifier() {
...
+            orgIdentifier, projectIdentifier, templateIdentifier);
+    return URLEncoder.encode(entityStringRaw, StandardCharsets.UTF_8.toString());
+  }
+
+  public String getEntityMetadataStringForTemplate(String templateIdentifier, String templateName)
+      throws UnsupportedEncodingException {
+    Map<String, String> metadataMap = ImmutableMap.<String, String>builder()
+                                          .put("templateIdentifier", templateIdentifier)
```
Comment (namanvermah) [inline 878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceHelper.java:151]: Please keep the entity metadata string as per how we discussed with @dwarren-dev

**Review by shivamnegi94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48336)* status: 19 file(s) to review. 2 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`template-service/service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1596 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWkbnXmCQy68n7GlEBx:-NWpGSlz-Wl81gDlchss:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/template-service/service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1596)):*
<details><summary><i>Previously, shalini-agr (Shalini Agrawal) wrote…</i></summary><blockquote>

let's rename this to OPA_EVALUATION_ACTION_SAVE
</blockquote></details>

Done.
___
*[`878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceService.java` line 19 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWka_CFEZdMC6C0Crr2:-NWpFcx93d7BKXXjxKP9:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceService.java#L19)):*
<details><summary><i>Previously, shalini-agr (Shalini Agrawal) wrote…</i></summary><blockquote>

if we are just planning to send template yaml here, can we rename this variable to templateJson rather than expanded json?
</blockquote></details>

Done.
___
*[`878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceHelper.java` line 154 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWkikBD4Y4uRUOKPf4q:-NWpFg89BGyWBYmOs4PE:bllhxcc) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceHelper.java#L154)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Why template service related code needs to be present here?
</blockquote></details>

It is specific to template and it is not related to template service, we need to create entityMetadata to pass to OPA client.
___
*[`878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java` line 141 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWkiwDR4o-uFi1_NQ_q:-NWpGDw2EG2mPweHEaiE:b-phqdm) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java#L141)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Lets try not to use template or pipeline keywords here
</blockquote></details>

But it is specific to template

File: template-service/service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java [hunk 3]
```diff
@@ -166,12 +166,23 @@ public ResponseDTO<TemplateWrapperResponseDTO> create(@NotNull String accountId,
           "created", templateEntity.getIdentifier(), gitEntityCreateInfo.getCommitMsg());
     }
 
+    GovernanceMetadata governanceMetadata = templateService.validateGovernanceRulesAndThrowExceptionIfDenied(templateEntity);
```
Comment (namanvermah) [inline template-service/service/src/main/java/io/harness/template/resources/NGTemplateResourceImpl.java:169]: why this method name if we are not adding any exception throwing?

**Review by namanvermah — APPROVED** [review]

**Review by shivamnegi94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48336)* status: 26 file(s) to review. 2 change requests. 3. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java` line 104 at r2](https://reviewable.io/reviews/harness/harness-core/48336#-NWkb7vwC47G65NseOLf:-NWq8YcT3zhMezYsqzb1:b-rxyxft) ([raw file](https://github.com/harness/harness-core/blob/384cff83abdf5d49f00c4f7cbff167065f65e3ea/878-ng-common-utilities/src/main/java/io/harness/engine/GovernanceServiceImpl.java#L104)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

+1
</blockquote></details>

We need to refactor the EntityString and EntityMetadata for pipeline policy method which will going to handle later.
@namanvermah Suggested to have separate method for template now and we can refactor it later once we fix the pipeline code.

**Pull request closed without merging**